File: 270. Closest Binary Search Tree Value/270.go
// 270. Closest Binary Search Tree Value solution in GoLang
// https://leetcode.com/problems/closest-binary-search-tree-value/description/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

package leetcode

func closestValue(root *TreeNode, target float64) int {

	closest := root.Val

	for root != nil {
		if math.Abs(float64(root.Val) - target) < math.Abs(float64(closest) - target) {
			closest = root.Val
		}

		if target < float64(root.Val) {
			root = root.Left
		} else {
			root = root.Right
		}
	}

	return closest
}

File: 270. Closest Binary Search Tree Value/270.js
// 270. Closest Binary Search Tree Value solution in JavaScript
// https://leetcode.com/problems/closest-binary-search-tree-value/description/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} target
 * @return {number}
 */
var closestValue = function(root, target) {
    let val = root.val;
    let closest = root.val;

    while (root != null) {
        val = root.val;

        closest = Math.abs(val - target) < Math.abs(closest - target) ? val : closest;

        root = target < root.val ? root.left : root.right;
    }

    return closest;
};